### PR TITLE
Necessary API Changed for Cpp Integration

### PR DESCRIPTION
**Pull request by ruixi1994**
## Proposed changes
Necessary API Changed for Cpp Integration

Describe what this PR is trying to do.

## Type of change

- [ ] Feature work
- [ ] Bug fix
- [ ] Documentation
- [x] Engineering change
- [ ] Test
- [ ] Logging/Telemetry

## Risk

- [ ] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [ ] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [x] Small – No issues are expected. (Example: Very small bug fixes, string changes, or configuration settings changes)

## Additional information
This change is to add missing component to integrate embeded webview implementation into Cpp. The changes in CPP repo for integration was made [there] (https://github.com/AzureAD/microsoft-authentication-library-for-cpp/pull/1038)

### CODE REVIEW

File: IdentityCore/src/oauth2/MSIDWebviewFactory.m [hunk 1]
```diff
@@ -105,6 +105,9 @@ @implementation MSIDWebviewFactory
                                                              endURL:[NSURL URLWithString:configuration.endRedirectUrl]
                                                             webview:webview
                                                       customHeaders:configuration.customHeaders
+#if TARGET_OS_OSX
+                                                       customWindow:NSZeroRect
```
Comment (oldalton) [inline IdentityCore/src/oauth2/MSIDWebviewFactory.m:109]: can we have it as an OS X only property? Initializers with conditional parameters are quite difficult to read/maintain.

File: IdentityCore/src/webview/MSIDWebviewAuthorization.m [hunk 2]
```diff
@@ -149,7 +149,7 @@ + (void)cancelCurrentSession
     {
         if (s_currentSession)
         {
-            [s_currentSession.webviewController cancel];
+            [s_currentSession.webviewController cancelProgramatically];
```
Comment (oldalton) [inline IdentityCore/src/webview/MSIDWebviewAuthorization.m:152]: nit: should be double m

File: IdentityCore/src/webview/systemWebview/session/MSIDSystemWebviewController.m [hunk 3]
```diff
@@ -145,20 +145,32 @@ - (void)startWithCompletionHandler:(MSIDWebUICompletionHandler)completionHandler
...
-    [self.session cancel];
+    [self.session cancelProgramatically];
     
+    [self notifyEndWebAuthWithURL:nil error:error];
+    self.completionHandler(nil, error);
+}
+
+- (void)cancelProgramatically
+{
+    MSID_LOG_WITH_CTX(MSIDLogLevelInfo, self.context, @"Authorization session was cancelled programatically");
```
Comment (oldalton) [inline IdentityCore/src/webview/systemWebview/session/MSIDSystemWebviewController.m:161]: should we also end telemetry event when cancelling programmatically?

File: IdentityCore/src/oauth2/MSIDWebviewFactory.m (hunk 1 above)
Comment (ruixi1994) [reply IdentityCore/src/oauth2/MSIDWebviewFactory.m:109]: Will make the change. Thanks!

Comment (oldalton) [reply IdentityCore/src/oauth2/MSIDWebviewFactory.m]: Not seeing the change for MSIDAADOAuthEmbeddedWebviewController. 
Regarding setting it in initializer... maybe you can pass a bag of properties? Like an object with all the parameters in the initializer. @antrix1989, any other ideas?

Comment (ruixi1994) [reply IdentityCore/src/oauth2/MSIDWebviewFactory.m]: Sorry missed the necessary change in that file. Will take the way of adding an object and make the change. Thanks for the advice!

File: IdentityCore/src/webview/embeddedWebview/ui/MSIDWebViewPlatformParams.h [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+#import <Foundation/Foundation.h>
+
+@interface MSIDWebViewPlatformParams : NSObject
+
+#if TARGET_OS_OSX
+
+@property (nonatomic, readonly) NSRect customWindowRect;
+
+-(id)initWithCoustomWindowRect: (NSRect)customWindowRect;
```
Comment (antrix1989) [inline IdentityCore/src/webview/embeddedWebview/ui/MSIDWebViewPlatformParams.h:33]: nit: remove space after ":".

File: IdentityCore/src/webview/embeddedWebview/ui/MSIDWebviewUIController.h [hunk 5]
```diff
@@ -37,13 +38,17 @@ NSWindowController
 @property (nonatomic) id<MSIDRequestContext> context;
 @property (nonatomic) BOOL loading;
 @property (nonatomic) BOOL complete;
+@property (nonatomic) MSIDWebViewPlatformParams *platformParams;
```
Comment (antrix1989) [inline IdentityCore/src/webview/embeddedWebview/ui/MSIDWebviewUIController.h:41]: Should platform params be "readonly"?

File: IdentityCore/src/webview/embeddedWebview/ui/mac/MSIDWebViewPlatformParams.m [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import <Foundation/Foundation.h>
+#import "MSIDWebViewPlatformParams.h"
+
+@implementation MSIDWebViewPlatformParams
+
+-(id)initWithCoustomWindowRect: (NSRect)customWindowRect {
```
Comment (antrix1989) [inline IdentityCore/src/webview/embeddedWebview/ui/mac/MSIDWebViewPlatformParams.m:29]: nit: remove space after ":" and move "{" to new line.

File: IdentityCore/src/webview/embeddedWebview/ui/mac/MSIDWebviewUIController.m [hunk 7]
```diff
@@ -169,7 +189,16 @@ - (void)windowWillClose:(__unused NSNotification *)notification
...
 - (NSProgressIndicator *)prepareLoadingIndicator
 {
-    NSProgressIndicator *loadingIndicator = [[NSProgressIndicator alloc] initWithFrame:NSMakeRect(DEFAULT_WINDOW_WIDTH / 2 - 16, DEFAULT_WINDOW_HEIGHT / 2 - 16, 32, 32)];
+    CGFloat windowWidth = DEFAULT_WINDOW_WIDTH;
+    CGFloat windowHeight = DEFAULT_WINDOW_HEIGHT;
+    if (_platformParams && !NSIsEmptyRect([_platformParams customWindowRect]))
+    {
+        NSRect window = [_platformParams customWindowRect];
+        windowWidth = window.size.width;
+        windowHeight =window.size.height;
```
Comment (antrix1989) [inline IdentityCore/src/webview/embeddedWebview/ui/mac/MSIDWebviewUIController.m:198]: nit: space after "=".

File: IdentityCore/src/webview/embeddedWebview/ui/MSIDWebviewUIController.h (hunk 5 above)
Comment (ruixi1994) [reply IdentityCore/src/webview/embeddedWebview/ui/MSIDWebviewUIController.h]: Nice Catch! Thanks!

File: IdentityCore/src/oauth2/MSIDWebviewFactory.m (hunk 1 above)
Comment (ruixi1994) [reply IdentityCore/src/oauth2/MSIDWebviewFactory.m]: Addressed

File: IdentityCore/src/webview/systemWebview/session/MSIDSystemWebviewController.m [hunk 8]
```diff
@@ -145,20 +145,32 @@ - (void)startWithCompletionHandler:(MSIDWebUICompletionHandler)completionHandler
...
     NSError *error = MSIDCreateError(MSIDErrorDomain,
                                      MSIDErrorSessionCanceledProgrammatically,
                                      @"Authorization session was cancelled programatically.", nil, nil, nil, self.context.correlationId, nil, YES);
-    
-    [self notifyEndWebAuthWithURL:nil error:error];
-    self.completionHandler(nil, error);
+    [self cancel:error];
+}
+
+- (void)userCancel {
```
Comment (antrix1989) [inline IdentityCore/src/webview/systemWebview/session/MSIDSystemWebviewController.m:168]: nit: new line

**Pull request merged**